### ISSUE TITLE
Dynamic light is flickering possibly a bug with BVH+Fog interaction

### ISSUE THREAD

**Issue by filipworkson** [godotengine/godot#46353]
**Godot version:**

3.2.4 rc3

**OS/device including version:**

NVIDIA 1650 Super
3.2.4 RC 3
GLES2

**Issue description:**

Dynamic light is now flickering in intensity in my game. 
I don't remember if previous RC 2 had this but it definitely didn't happen in 3.2.3 .
If I turn BVH off the flicker goes away. I'm fairly confident is related to BVH changes.  

**Edit: Further testing reveals is related to BVH + Fog option turned on. It doesn't happen with BVH off or with BVH on + Fog off and it doesn't happen when you are very close to the light only when you are quite a bit away. Might also be related with the Fog option `fog_transmit_enabled` ** 

![a](https://user-images.githubusercontent.com/36826319/108880898-bbd6dc80-75d0-11eb-9174-52297d085e55.gif)

Looks like my game is ready for Christmas 🎄 🤣  

**Steps to reproduce:**

See minimal reproduction project. Is a complex interaction between multiple components. bvh + animated point meshes + fog + fog_transmit_enabled + dynamic light with cast shadow enabled.

**Minimal reproduction project:**

[bvh_test.zip](https://github.com/godotengine/godot/files/6031132/bvh_test.zip)

**Comment by Calinou** [comment]
@filipworkson Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by pouleyKetchoupp** [comment]
Please upload a minimal reproduction project to help contributors investigate the issue.

**Comment by filipworkson** [comment]
@lawnjelly  I'm sorry it took me so long to provide a test project. I thought it might  be difficult to reduce it down but it was simpler than I thought!

Minimal test project:
[bvh_test.zip](https://github.com/godotengine/godot/files/6031080/bvh_test.zip)

- I had to strip off all the scripts so you can't walk around but you can manually move the player around so hopefully you should see the visual glitches. 
- The viewport used to resize automatically to screen resolution but without the script is now set to a fixed ~700 x ~400 just to see enough details. 
- I also kept an item in which is a custom billboarded shader to test light updates when meshes are disabled and enabled from my previous post.

Based on my tests it seems to be a bizarre interaction between animated point meshes with custom shader applied + bvh turned on + dynamic light + fog + fog_transmit_enabled turned on. If you disable player body the glitch goes away. I think bvh updates when meshes move near a dynamic light and it inadvertently affects light transmit option for fog.

**Comment by lawnjelly** [comment]
Will take a look. :+1:

**Comment by lawnjelly** [comment]
Ok I've managed to reduce it down to a minimum reproduction project. For future reference it really helps if you can do this, because programmer time is limited, it gets issues fixed much faster.
[bvh_test_filip.zip](https://github.com/godotengine/godot/files/6031647/bvh_test_filip.zip)

Not absolutely sure what is causing it yet. Despite the flickering, the pairing is not pairing and unpairing instances (lights etc) so I can't discount the possibility that the flickering is coming from elsewhere, but being triggered by the BVH. I'll do some more debugging output and see if I can work out what is going on.

**Comment by lawnjelly** [comment]
I've finally tracked this down (to the general area). As I was suspecting, it's not actually a bug in the BVH.

There is a difference between the BVH and the octree, the BVH is continuously doing an incremental optimize of the tree structure each update (frame). Items are taken out and reinserted in the tree, to see if they find a better position .. in many cases they don't. But the upshot is, that the order of items in the tree changes, and although the results of a `convex_cull` test are the same (in terms of objects hit), the order of the results can change.

There is something in the fog (or lighting, not sure which yet) in GLES2 which is order dependent. This is very probably a bug or oversight. It may well have already exhibited with octree, but been less visually noticable. It is probably either it is relying on the lights being reported in the cull hits before the meshes, or vice versa, or the overall order.

The cull results in my minimum reproduction project are four object, 2 lights and 2 meshes (the background, and the leg). The result order in the octree is:
```
	-1, 0, -1 - 2.00001, 6.00001, 2
	-4, -4, -4 - 8, 8, 8
	-0.0625, -0.546875, -0.078125 - 0.15625, 0.5626, 0.15625
	-4, -4, -4 - 8, 8, 8
``` 
I assume mesh, light, mesh, light.

With the BVH the order is:
```
	-0.0625, -0.546875, -0.078125 - 0.15625, 0.5626, 0.15625
	-1, 0, -1 - 2.00001, 6.00001, 2
	-4, -4, -4 - 8, 8, 8
	-4, -4, -4 - 8, 8, 8
```
and
```
	-1, 0, -1 - 2.00001, 6.00001, 2
	-0.0625, -0.546875, -0.078125 - 0.15625, 0.5626, 0.15625
	-4, -4, -4 - 8, 8, 8
	-4, -4, -4 - 8, 8, 8
```
mesh, mesh, light, light, flickering between the two.

If I turn off the BVH frame update (optimizes) the flickering disappears.
This may be due to some additive nature of the lighting being order dependent in GLES2.

**Comment by lawnjelly** [comment]
Looks like it is the order of the lights in the cull results in the problem rather than the mesh ordering. This could make it easier to (bodge) fix, as although there could be a large number of meshes the number of lights should be limited, so sorting them each frame is far less likely to be a performance problem.

I'm of the opinion that fundamentally the problem is down to the fog / lighting algorithm being order dependent, which is kinda gross and may make it impossible to provide entirely repeatable lighting - but maybe that's the only option with multipass lighting. And we can bodge around this issue by sorting the lights. Sorting simply on the light pointer seems to stabilize the flickering in the test case (and may help when the octree is used too, although this may not have been noticed before). Maybe there's something else we can sort on that makes more sense that address in memory (which will be largely random from run to run).

This light sorting is a good approach because I'm loathe to put hacks into the BVH to try and stabilize ordering, and we should be able to make light sorting GLES2 only.

**Comment by filipworkson** [comment]
@lawnjelly 
I am sorry you always have to deal with my confusing and obscure bug reports. I been trying my best to help out this time with a test project and for future I will minimize it as you have specified. 

Now I don't know the internals but I did my own set of testing and in the following situations the problem is no longer present:
1. Hide the moving leg or turn off moving animation. The movement of the leg  as it gets closer/farther from the camera  disrupts the order of meshes.  If you turn off the movement and make it static problem is no longer present. 
2. Turn off `fog_transmit_enabled`.  I think it might be an issue with how fog transmit happens. Fog transmit will allow light to leak though fog and I guess it might be using octree order before the convex cull  where as dynamic light might be using bvh?
![image](https://user-images.githubusercontent.com/36826319/109008223-10369680-767b-11eb-8a06-1da8281cec49.png)
3. Turning off BVH. If you say bvh gives slightly different results from octree calculations then it might not be from bvh at all. It might be some difference of calculations. 

My guess is the fog transmit light is the culprit. It might be using octree values before convex cull so the resulting intensity is slightly off for 1 frame. But you have multiple dynamic lights the effect is a kind of cascade where each light slightly dims for a brief moment probably based on the actual tree order.   Looks like a christmas tree light where each light dims in a sequence then loops.

**Comment by Calinou** [comment]
@filipworkson Note that fog transmittance hasn't been reimplemented in 4.0. If you plan on ugprading, it may be a good idea to stop using it.

**Comment by filipworkson** [comment]
@Calinou  @lawnjelly  Yes I think the best solution is just to turn off `fog_transmit_enabled` option when using bvh. Which is what I am doing right now.

This bug seems pretty obscure. What is interesting , however, is that I don't remember it happening in 3.2.4 rc1 or rc2 though it might be possibly have happened just didn't notice it until now.

**Comment by lawnjelly** [comment]
Yes it is the `fog transmit`, however it is possible that there are other undiscovered GLES2 lighting artifacts that may occur if they rely on light ordering. I also found hiding the leg had an effect, but it probably indirect by making it less likely the lights will reorder.

Anyway I'll know soon enough when I've figured out a good place to put in light sorting. But it works in the test case, and it makes a lot of sense in terms of mechanism (if it renders all objects with light 1, then all objects with light 2 etc).
